rle_array/_slicing.py

```diff
@@ -0,0 +1,395 @@
+
+The issue with Python slicing are:
+
+- ``slice`` type:
+  - the types in ``slice`` are completely unchecked (can even be a string or any user-provided type)
+  - the consistency of the values in ``slice`` are unchecked
+  - there is not information about the container size (which makes consistency checks more complicated)
+
+- ``slice.step`` value:
+  - ``step`` has the implicit default 1
+  - there can be forward and backward slices depending on the ``step`` value
+  - there can be step sizes which are not modulo 1
+
+- ``slice.start`` and ``slice.stop`` values:
+  - the implicit defaults of ``start`` and ``stop`` depend on ``step`` (is it positive or negative?)
+  - ``start`` and ``stop`` can be negative (aka "from the end")
+  - ``start`` and ``stop`` can over/underflow the container
+
+We do not want to deal with all these edge cases in every code snipped that deals with slicing, so we introduce
+:class:`NormalizedSlice` that solves the issue in a central place.
+"""
+from typing import Optional
+
+import numpy as np
+
+
+class NormalizedSlice:
+    """
+    A normalized slice.
+
+    .. important::
+
+        Do not try to construct this class by hand. Use :func:`NormalizedSlice.from_slice` instead!
+
+    Parameters
+    ----------
+    start
+        First absolute index in the container (inclusive start). Always positive.
+    stop
+        Last absolute index not being part of the slice (exclusive end). Counted from the container start. Can be
+        negative for refersed slices (aka ``step < 0``). Must be normalized so that ``abs(stop - start) % step == 0``.
+        For forward slices (``step > 0``), this must be greater than ``start``. For backward slices (``step < 0``) this
+        must be less than ``start``. For empty slices (``start = stop``), ``start``, ``stop`` and ``step`` have the
+        fixed values 0, 0 and 1.
+    step
+        Step size. Must not be ``0``.
+    container_length
+        Size of the container this slice applies to. Must not be negative. For empty containers
+        (``container_length = 0``), ``start``, ``stop`` and ``step`` have the fixed values 0, 0 and 1.
+    """
+
+    def __init__(self, start: int, stop: int, step: int, container_length: int):
+        if not isinstance(start, int):
+            raise TypeError(f"start must be int but is {type(start).__name__}")
+        if not isinstance(stop, int):
+            raise TypeError(f"stop must be int but is {type(stop).__name__}")
+        if not isinstance(step, int):
+            raise TypeError(f"step must be int but is {type(step).__name__}")
+        if not isinstance(container_length, int):
+            raise TypeError(
+                f"container_length must be int but is {type(container_length).__name__}"
+            )
+
+        self._start = start
+        self._stop = stop
+        self._step = step
+        self._container_length = container_length
+
+        self._verify()
+
+    def _verify(self) -> None:
```

Comment:
maybe call it `_verify_container_length` to make it obvious what the function actually does.

Reply:
it validates all parameters, not just the length.